designs/function-node-lifecycle/README.md

```diff
@@ -53,6 +53,50 @@ Function node declares its required NPM module list to runtime by using new API.
 
 Because current Node-RED implementation starts flow execution before message handler callback is ready, we may need to modify runtime execution model in order to make function nodes wait for receiving messages before required NPM modules become ready to use.
 
+#### Additional settings for external npm module in `settings.js`
+
+Add a top-level property named `externalModule` to `settings.js`.  It points to a object with following properties.
+
+| name | type   | required | description  |
+| :--- | :----- | :------- | :----------- |
+| mode | string | yes      | one of `"auto"`, `"auto-update"`, `"manual"`, or `"none"` (default: `"manual"`)<br/>- `auto`: automatically install requested NPM module if not installed.<br/>- `auto-update`: same as `auto` but with automatic update if newer version is available.<br/>- `manual`: users can manually install requested NPM module from user settings panel.<br/>- `none`: installation of external NPM modules is not allowd |
+| whiteList | array of string | no | List of regular expressions that matches module's NPM registry specification. External libraries that matches this list **can** be used in `Function` nodes. |
+| blackList | array of string | no | List of regular expressions that matches module's NPM registry specification. External libraries that matches this list **can not** be used in `Function` nodes. |
+
+`whiteList` precedes over `blackList`.
+
+```
+    // Example of settings.js
+    externalModules: {
+        mode: "manual",
+        whiteList: [
+            "^fs-ext$",       // allow fs-ext
+            "^qrcode@1.4.4$"  // allow qrcode version 1.4.4
+        ],
+        blackList: [
+            ".*"              // disallow others 
+        ]
+    },
+```
+If the `externalModule` property not exists, `none` mode and blackList of `".*"`, meaning not allow installation and use of external module, is expected as a default.
+
+#### New API for requesting NPM module
+
+- `RED.nodes.registerRequiredModules`(*module*)
+
+Requests Node-RED runtime to install NPM module *module*.  Returns a promise that resolves to required module object or rejects if installation failed.  NPM module installation is performed according to specification in `externalModule` property of `settiongs.js` described above.
+
+If `auto` mode is specified, execution of a function node setup or body that require NPM modules waits for completion of their installation.
+
+#### Manual installation of NPM module
+
+User Settings panel will be extended to have `Modules` tab.
```

Comment:
The `dynamicModuleList` concept I mentioned above could be used here as well. That means it would be able to list modules added to Function nodes but not yet deployed.

How does the editor know there is an update for a module? We can do that for the Palette Manager because we can download the full catalog. There is no equivalent catalog for npm.

I am concerned that users might get confused about what this tab is for - particularly users who are already familiar with their `.node-red/package.json` file. They may expect this list to include everything in their package.json file to help them manage it. I don't think we want to do that.

The more I think about it, I would like to discuss **not** adding this tab. Instead, the module list on the individual Function nodes could provide the same `install`/`check update` options. That makes more sense to me otherwise the user has to constantly switch between the Function node and this settings dialog to manage the modules. It would be better to be able to manage it in the Function node UI.

Given other nodes may want to provide similar functionality (such as ui_builder), this may need to be a common UI widget that can be reused... that is probably a thought for future once we understand this feature better.

Maybe not for the first release, but need to think about how the user can specify a version of the module to install - and understand what happens if two different Function nodes request two different versions (simply not possible to support that).

Reply:
One more comment - for users with Projects enabled,  the runtime should update the project's `package.json` file to include the modules nodes depend on. That will ensure the modules can be preinstalled when the project is deployed somewhere else.

There are other issues around dependency management for projects that already need fixing - but this would be another reason to get them fixed.

Reply:
The configuration panel of the original proposal has been merged into the configuration panel of the Function node. 
The original intent of global setting was to avoid version conflicts of a module as you point out, but I feel it would be better to be able to load multiple versions, so I changed it that way.

Reply:
@HiroyasuNishiyama in our recent conversation, you said you thought the module management UI should move back to the User Settings dialog and not under the individual nodes. Could you please explain that with reference to the discussion here where I explained why I thought it *should* be under the individual nodes?

>  but I feel it would be better to be able to load multiple versions

Allowing multiple versions of the same module to be installed would make the runtime code very complicated. Each node would have to get its very own `node_modules` directory somewhere on the file system, and for the `require` statement to point at the right directory. It would also mean if a module is used by multiple nodes, even at the same version, it would have to be installed separately for each module. Is that what you are proposing in this design?

I suggest we do not try to make this any more complicated then it already is for the first version. Start from a position that we can only install a single version of a module - and then think through what that means for the user experience. The Editor could do the work to ensure all of the Function nodes 'agree' on the version to be installed - or highlights to the user where there are incompatible versions strings.

We can always add more discrete version support in a future iteration.

Reply:
> @HiroyasuNishiyama in our recent conversation, you said you thought the module management UI should move back to the User Settings dialog and not under the individual nodes. Could you please explain that with reference to the discussion here where I explained why I thought it _should_ be under the individual nodes?

@knolleary let me summarize the currently proposed NPM installation methods.  
Could you please point out your thoughts and any differences in your assumptions.

In the following discussion, we assume SUBFLOW **S** contains a function node that requires external NPM module **X**.

![FuncNode-Ex0](https://user-images.githubusercontent.com/30289092/100723447-1c58e200-3405-11eb-97da-cf51574e78a4.png)

1. Function node within a flow (or non-imported SUBFLOW)

   The function node provides an interface for specifying required NPM modules and associated variable name visible in its JavaScript code.  

![FuncNode-IF0](https://user-images.githubusercontent.com/30289092/100723470-267ae080-3405-11eb-991e-441d607d72f7.png)

   The function node may also provides interface for install/update/uninstall NPM module feature or may provides editor main menu for it.  

2. Function node within imported SUBFLOW

   The imported SUBFLOW is treated as a pseudo node, so settings interface of the node contained in the SUBFLOW cannot be accessed. 

   a. Use same interface & mechanism of (1)

   ​	As mentioned above, in this case, we can not access interface of nodes within SUBFLOW.
   
   b. Use dependent module specification of `package.json`
   
   ​	The export mechanism of SUBFLOW scans the setting information of Function nodes within the SUBFLOW and outputs the information of the dependent module to `package.json`.  When installing the SUBFLOW module, the required modules are installed under the installed SUBFLOW node module (*NR*`/ node_modules /`*NPM*` / nodde_modules`).
   
![NPM-install00](https://user-images.githubusercontent.com/30289092/100723571-44e0dc00-3405-11eb-88ae-2be120126b3b.png)

(2)-b allows the imported NPM module similar to independent node.

(2)-a is consistent with (1) because they can share same internal API for manipulating NPM modules.  However, there is a problem that the imported SUBFLOW cannot access the interface of the internal Function node.  

This problem may be solved by moving NPM module install/update/unstall interface to main editor menu.

![NPM-UI-separation00](https://user-images.githubusercontent.com/30289092/100723517-34c8fc80-3405-11eb-827a-e4b610be6c29.png)

Separation of the NPM operation interface may complicate the UI as you pointed out, but it has the following additional advantages for (1) and (2)-a.

- If installation of NPM module fail, checking and fixing the problem can be done with single UI.
- If each node has an NPM operation interface, different versions can be specified.  With single NPM manipulation UI, it may be solved simply.



> > but I feel it would be better to be able to load multiple versions
> 
> Allowing multiple versions of the same module to be installed would make the runtime code very complicated. Each node would have to get its very own `node_modules` directory somewhere on the file system, and for the `require` statement to point at the right directory. It would also mean if a module is used by multiple nodes, even at the same version, it would have to be installed separately for each module. Is that what you are proposing in this design?

In the original design, different version of the same NPM modules can be specified in Function nodes settings panel.  I thought it was necessary to allow those installations to eliminate conflicts.
As you pointed out, it may be ignored in the first edition.  Or it may be handled with single UI as explained above.

Reply:
A user has no knowledge of the internals of a subflow that has been installed as an npm module. They install the subflow module and it appears just like any other node. If that subflow contains a Function node that has its own npm dependencies, then the subflow module's package.json must list those dependencies to ensure they get installed by npm when the subflow module is installed.

If the npm install of a dependency of the subflow module fails to install, then the whole subflow module will fail to install.

*The end-user is not expected to manage those dependencies themselves.*

Subflow modules are not a reason to change the UI design for the Function node module handling.

---

Focussing purely on the Function nodes themselves, I described my concerns over having it in User Settings in [this comment](https://github.com/node-red/designs/pull/31#discussion_r451850432). But I'm less sure of the right approach now.

I do know that managing multiple versions of an installed module is a **very hard** problem. I don't mean from a UI design, but from the runtime side. If you have a requirement to install multiple versions, then this design needs to outline how that will work in the runtime side before we do anything else.

If we restrict it to a single version, then having a central dialog, as you propose here, to manage it would make sense - there is *one* place to manage the modules so it is clear you can't have different versions in each node.

Reply:
@knolleary  I understand your point about the SUBFLOW NPM module.  For the SUBFLOW NPM module, I agree that the NPM module installation follows only the description in package.json.

Regarding installation of different version of NPM modules, I suppose  there are such use cases because different Function nodes may require different version of the same NPM module.  My original assumption was to use npm alias for handling this.  But as mentioned in the teleconference, I would like to suggest having only one NPM installation panel, and not allowing installation of different version of NPM module  in the initial version as your suggestion.


> The more I think about it, I would like to discuss **not** adding this tab. Instead, the module list on the individual Function nodes could provide the same `install`/`check update` options. That makes more sense to me otherwise the user has to constantly switch between the Function node and this settings dialog to manage the modules. It would be better to be able to manage it in the Function node UI.



To address this concern, let me suggest transitioning from the NPM module selection menu of Function node configuration palel (possibly as global config node) to NPM module management panel for installing new NPM module.  By sharing specification on this panel between Function nodes, I think we can manage NPM modules in one place instead of switching to settings menu.

By sharing specification on this panel between Function nodes, I think we can manage NPM modules in one place.

![NPM-module-mgmt](https://user-images.githubusercontent.com/30289092/101319649-a8f01e00-38a5-11eb-846b-74061fc3978d.png)


Please let me know what you think.